### PR TITLE
Gluon InstanceNorm and ReflectancePadding

### PR DESCRIPTION
**Pull request by zhanghang1989**
InstanceNorm and ReflectancePadding are important for generative models and style transfer.

### CODE REVIEW

**Comment by piiswrong** [comment]
Please add doc. see http://pytorch.org/docs/master/nn.html#torch.nn.InstanceNorm1d

File: python/mxnet/gluon/nn/basic_layers.py [hunk 1]
```diff
@@ -442,3 +442,87 @@ def hybrid_forward(self, F, x):
...
+    carried out per example (instance), not over a batch. Second, the
+    same normalization is applied both at test and train time. This
+    operation is also known as `contrast normalization`.
+    If the input data is of shape [batch, channel, spacial_dim1, spacial_dim2, ...],
+    `gamma` and `beta` parameters must be vectors of shape [channel].
+    This implementation is based on paper:
+    .. [1] Instance Normalization: The Missing Ingredient for Fast Stylization,
+       D. Ulyanov, A. Vedaldi, V. Lempitsky, 2016 (arXiv:1607.08022v2).
+    Examples::
+      // Input of shape (2,1,2)
```
Comment (szha) [inline python/mxnet/gluon/nn/basic_layers.py]: add an empty line after ::
http://www.sphinx-doc.org/en/stable/rest.html#source-code

File: python/mxnet/gluon/nn/basic_layers.py [hunk 2]
```diff
@@ -442,3 +442,90 @@ def hybrid_forward(self, F, x):
...
+
+class InstanceNorm(HybridBlock):
+    """
+    Applies instance normalization to the n-dimensional input array.
+    This operator takes an n-dimensional input array where (n>2) and normalizes 
+    the input using the following formula:
+    .. math::
+
+    out = \frac{x - mean[data]}{ \sqrt{Var[data]} + \epsilon} * gamma + beta
+    This layer is similar to batch normalization layer (`BatchNorm`)
```
Comment (szha) [inline python/mxnet/gluon/nn/basic_layers.py]: A new line here too, since the block of source code is supposed to be surrounded by empty lines.

Comment (zhanghang1989) [reply python/mxnet/gluon/nn/basic_layers.py]: Does it look okay now? I am not familiar with rst

**Comment by zhanghang1989** [comment]
Let me know if it is ready to merge. I will update another PR accordingly https://github.com/apache/incubator-mxnet/pull/7579. @piiswrong @szha  Thanks!

**Comment by piiswrong** [comment]
Front end interface APIs need to conform to the highest standards.

Please look at other APIs and fix the formatting.

File: python/mxnet/gluon/nn/basic_layers.py [hunk 3]
```diff
@@ -442,3 +442,92 @@ def hybrid_forward(self, F, x):
...
+    with two differences: first, the normalization is
+    carried out per example (instance), not over a batch. Second, the
+    same normalization is applied both at test and train time. This
+    operation is also known as `contrast normalization`.
+    If the input data is of shape [batch, channel, spacial_dim1, spacial_dim2, ...],
+    `gamma` and `beta` parameters must be vectors of shape [channel].
+    This implementation is based on paper:
+    .. [1] Instance Normalization: The Missing Ingredient for Fast Stylization,
+       D. Ulyanov, A. Vedaldi, V. Lempitsky, 2016 (arXiv:1607.08022v2).
+    Examples::
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/basic_layers.py]:
```
Examples
---------
>>> ...
```

File: python/mxnet/gluon/nn/basic_layers.py [hunk 4]
```diff
@@ -442,3 +442,92 @@ def hybrid_forward(self, F, x):
 
     def __repr__(self):
         return self.__class__.__name__
+
+
+class InstanceNorm(HybridBlock):
+    """
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/basic_layers.py]: missing argument documentation

File: python/mxnet/gluon/nn/basic_layers.py [hunk 5]
```diff
@@ -442,3 +442,92 @@ def hybrid_forward(self, F, x):
...
+        s += ')'
+        return s.format(name=self.__class__.__name__,
+                        content=', '.join(['='.join([k, v.__repr__()])
+                                           for k, v in self._kwargs.items()]))
+
+
+class ReflectionPad(HybridBlock):
+    """Pads the input tensor using the reflection of the input boundary.
+
+    Args:
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/basic_layers.py]: See other APIs for formatting.

File: python/mxnet/gluon/nn/basic_layers.py [hunk 6]
```diff
@@ -442,3 +442,92 @@ def hybrid_forward(self, F, x):
...
+        s = '{name}({content}'
+        if hasattr(self, 'in_channels'):
+            s += ', in_channels={0}'.format(self.in_channels)
+        s += ')'
+        return s.format(name=self.__class__.__name__,
+                        content=', '.join(['='.join([k, v.__repr__()])
+                                           for k, v in self._kwargs.items()]))
+
+
+class ReflectionPad(HybridBlock):
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/basic_layers.py]: Should be named ReflectionPad2D and put in convolution layers

File: python/mxnet/gluon/nn/basic_layers.py [hunk 7]
```diff
@@ -442,3 +442,92 @@ def hybrid_forward(self, F, x):
...
+          :math:`H_{out} = H_{in} + paddingTop + paddingBottom`
+          :math:`W_{out} = W_{in} + paddingLeft + paddingRight`
+
+    Examples::
+
+        m = nn.ReflectionPad(3)
+        input = mx.nd.random_normal(shape=(16, 3, 224, 224))
+        output = m(input)
+    """
+    def __init__(self, pad_width=None, **kwargs):
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/basic_layers.py]: what does a None default mean?

File: python/mxnet/gluon/nn/basic_layers.py [hunk 8]
```diff
@@ -442,3 +442,92 @@ def hybrid_forward(self, F, x):
...
+        super(ReflectionPad(, self).__init__(**kwargs)
+        self.pad_width = pad_width
```
Comment (piiswrong) [inline python/mxnet/gluon/nn/basic_layers.py]: _pad_width

File: python/mxnet/gluon/nn/basic_layers.py (hunk 4 above)
Comment (zhanghang1989) [reply python/mxnet/gluon/nn/basic_layers.py]: Thanks! The args are added here: https://github.com/apache/incubator-mxnet/pull/7570/files#diff-06285baf97b6456dd23bdb99f6e239b6R461

File: python/mxnet/gluon/nn/basic_layers.py [hunk 9]
```diff
@@ -453,31 +448,65 @@ class InstanceNorm(HybridBlock):
     .. math::
 
       out = \frac{x - mean[data]}{ \sqrt{Var[data]} + \epsilon} * gamma + beta
-
+
```
Comment (szha) [inline python/mxnet/gluon/nn/basic_layers.py]: Remove the bleeding whitespace.

**Comment by zhanghang1989** [comment]
Hey folks, I was busy with my thesis defense last week. Any further feedbacks for this PR?

**Comment by szha** [comment]
I suppose this means you passed. Congrats :)

Could you do a rebase onto the latest master to let the tests run?

**Comment by zhanghang1989** [comment]
Thanks! I have just rebased. Let me know if I am doing something wrong.

**Comment by zhanghang1989** [comment]
Got messed up ... Creating a new PR

**Pull request closed without merging**